### PR TITLE
Do not allow structured binding initializer referencing introduced name

### PR DESCRIPTION
**Pull request by i-garrison**
If name lookup ends up with a variable, additionally check if node being resolved is part of the structured binding initializer introducing found variable. If this is the case, produce problem binding and report structured binding declaration error via codan.

This change also prevents infinite recursion trying to resolve auto type of introduced variable while evaluating such problematic initializer.

### CODE REVIEW

**Comment by jonahgraham** [comment]
Added call for reviewers on [cdt-dev](https://www.eclipse.org/lists/cdt-dev/msg35425.html). If none are forthcoming we'll need a plan-B (me learning this area?)

File: core/org.eclipse.cdt.core/parser/org/eclipse/cdt/internal/core/dom/parser/cpp/semantics/CPPSemantics.java [hunk 1]
```diff
@@ -409,6 +409,24 @@ private static IBinding postResolution(IBinding binding, LookupData data) {
 			}
 		}
 
+		// If any name in the initializer refers to introduced names then declaration is ill-formed.
+		if (binding instanceof ICPPInternalBinding && ((ICPPInternalBinding) binding).getDefinition() != null) {
+			IASTNode definitionParent = ((ICPPInternalBinding) binding).getDefinition().getParent();
+			if (definitionParent instanceof ICPPASTStructuredBindingDeclaration) {
+				ICPPASTStructuredBindingDeclaration declaration = (ICPPASTStructuredBindingDeclaration) definitionParent;
```
Comment (ddscharfe) [inline core/org.eclipse.cdt.core/parser/org/eclipse/cdt/internal/core/dom/parser/cpp/semantics/CPPSemantics.java:416]: Just my personal taste, for this kind of if statements pattern matching comes in handy
```suggestion
		if (binding instanceof ICPPInternalBinding internalBinding && internalBinding.getDefinition() != null
				&& internalBinding.getDefinition()
						.getParent() instanceof ICPPASTStructuredBindingDeclaration declaration) {
```

**Review by ddscharfe — APPROVED** [review]
Looks good to me.
I could reproduce the error, tested that it is now covered by the new check and couldn't find a structured binding which isn't detected.

**Comment by i-garrison** [comment]
Hi @jonahgraham could you please approve workflows for this PR? I do not have any merged PRs yet.

**Review by jonahgraham — APPROVED** [review]
I rebased the change and correct `@since` tags. Once CI approves this is ready to be merged.

**Pull request merged**